kubeprod/pkg/eks/platform.go

```diff
@@ -0,0 +1,312 @@
+
+	aupInput := &iam.AttachUserPolicyInput{
+		PolicyArn: policyArn,
+		UserName:  aws.String(userName),
+	}
+	svc := iam.New(conf.getAwsSession())
+	_, err := svc.AttachUserPolicy(aupInput)
+	if err != nil {
+		log.Panicf("Error attaching policy %s to user %s\n%v", *policyArn, userName, err)
+	} else {
+		log.Info("Attached IAM policy for External DNS integration")
+	}
+}
+
+// Creates a new user (or reuses the existing one) in AWS to allow
+// for integration between External DNS and a hosted Route53 zone.
+// The user is named like "bbkpr-${dnsZone}" and will get an IAM
+// policy attached to it which limits R/W to the hosted Route53 zone
+// to be used by BKPR and R/O for any other zones. The IAM policy
+// will be created if necessary.
+func (conf *Config) createAwsUser() (string, string) {
+	userName := fmt.Sprintf("bkpr-%s", conf.DNSZone)
+
+	// Create an AWS user
+	userInput := &iam.CreateUserInput{
+		UserName: aws.String(userName),
+		Tags: []*iam.Tag{
+			{
+				Key:   aws.String("created_by"),
+				Value: aws.String("bkpr"),
+			},
+		},
+	}
+
+	svc := iam.New(conf.getAwsSession())
+	_, err := svc.CreateUser(userInput)
+	if err != nil {
+		log.Warning("Re-using existing AWS user for External DNS integration: ", userName)
+	} else {
+		log.Infof("Created AKS user: %s", userName)
+	}
+
+	conf.attachUserPolicy()
+
+	// Create/Add an Access Key
+	akInput := &iam.CreateAccessKeyInput{
+		UserName: aws.String(userName),
+	}
+	ak, err := svc.CreateAccessKey(akInput)
+	if err != nil {
+		log.Panicf("Cannot create AWS access key for External DNS integration\n%v", err)
+	}
+	return *ak.AccessKey.AccessKeyId, *ak.AccessKey.SecretAccessKey
+}
+
+// Configuration for integration between External DNS and AWS.
+func (conf *Config) setUpExternalDNS() error {
+	log.Info("Setting up configuration for External DNS")
+	flags := conf.flags
+
+	if conf.ExternalDNS.AWSAccessKeyID == "" {
+		AWSAccessKeyID, err := flags.GetString(flagAWSAccessKeyID)
+		if err != nil {
+			return err
+		}
+		conf.ExternalDNS.AWSAccessKeyID = AWSAccessKeyID
+	}
+	if conf.ExternalDNS.AWSSecretAccessKey == "" {
+		AWSSecretAccessKey, err := flags.GetString(flagAWSSecretAccessKey)
+		if err != nil {
+			return err
+		}
+		conf.ExternalDNS.AWSSecretAccessKey = AWSSecretAccessKey
+	}
+
+	// At this point, if the AWS secret is still empty, try to create an AWS
+	// access key for a user named "bkpr.${dnsZone}"
+	if conf.ExternalDNS.AWSAccessKeyID == "" || conf.ExternalDNS.AWSSecretAccessKey == "" {
+		conf.ExternalDNS.AWSAccessKeyID, conf.ExternalDNS.AWSSecretAccessKey = conf.createAwsUser()
+	}
+	return nil
+}
+
+// Generate platform configuration
+func (conf *Config) Generate(ctx context.Context) error {
+	flags := conf.flags
+
+	if conf.DNSZone == "" {
+		domain, err := flags.GetString(flagDNSSuffix)
+		if err != nil {
+			return err
+		}
+		conf.DNSZone = domain
+	}
+
+	if conf.DNSZone != "" {
+		//
+		// External DNS setup
+		//
+		err := conf.setUpExternalDNS()
```

Comment:
I think this will increase the amount of code and complexity substantially. Just to handle time-outs I have changed the code that creates the AWS session to provide an explicit time-out of 30 seconds for all HTTP requests to the AWS API, which is about 5 additional lines of code.